pkg/bootstrap/leaderelector/kube/factory.go

```diff
@@ -55,39 +61,91 @@ func (f *kubeElectionFactory) StartElection(ctx context.Context, config *leadere
 		return identity, err
 	}
 
-	l, err := k8sleaderelection.NewLeaderElector(
-		k8sleaderelection.LeaderElectionConfig{
-			Lock:          resourceLock,
-			LeaseDuration: getLeaseDuration(),
-			RenewDeadline: getRenewPeriod(),
-			RetryPeriod:   getRetryPeriod(),
-			Callbacks: k8sleaderelection.LeaderCallbacks{
-				OnStartedLeading: func(callbackCtx context.Context) {
-					contextutils.LoggerFrom(callbackCtx).Debug("Started Leading")
-					close(elected)
-					config.OnStartedLeading(callbackCtx)
-				},
-				OnStoppedLeading: func() {
-					contextutils.LoggerFrom(ctx).Error("Stopped Leading")
-					config.OnStoppedLeading()
-				},
-				OnNewLeader: func(identity string) {
-					contextutils.LoggerFrom(ctx).Debugf("New Leader Elected with Identity: %s", identity)
-					config.OnNewLeader(identity)
+	var counter atomic.Uint32
+	var justFailed = false
+	var dontDie func()
+
+	newLeaderElector := func() (*k8sleaderelection.LeaderElector, error) {
+
+		dieIfUnrecoverable := func(ctx context.Context) {
```

Comment:
some documentation here would be helpful. i gave it a first try, but i'm not sure if it's entirely accurate

```suggestion
                // this function causes gloo to exit after a specified amount of time of trying to recover after leader election failure. we do this to safeguard against ... (to be completed)
		dieIfUnrecoverable := func(ctx context.Context) {
```

Reply:
Thanks, added comments around this function